src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -50,153 +51,137 @@ case class InferReadWriteAnnotation(t: String, tID: TransID)
 object InferReadWritePass extends Pass {
   def name = "Infer ReadWrite Ports"
 
-  def inferReadWrite(m: Module) = {
-    import WrappedExpression.we
-    val connects = HashMap[String, Expression]()
-    val repl = HashMap[String, Expression]()
-    val stmts = ArrayBuffer[Statement]()
-    val zero = we(UIntLiteral(0, IntWidth(1)))
-    val one = we(UIntLiteral(1, IntWidth(1)))
+  type Netlist = collection.mutable.HashMap[String, Expression]
+  type Statements = collection.mutable.ArrayBuffer[Statement]
+  type PortSet = collection.mutable.HashSet[String]
+
+  private implicit def toString(e: Expression) = e.serialize
 
-    // find all wire connections
-    def analyze(s: Statement): Unit = s match {
-      case s: Connect  =>
-        connects(s.loc.serialize) = s.expr
-      case s: PartialConnect =>
-        connects(s.loc.serialize) = s.expr
-      case s: DefNode =>
-        connects(s.name) = s.value
-      case s: Block =>
-        s.stmts foreach analyze
-      case _ =>
+  def getProductTerms(connects: Connects)(e: Expression): Seq[Expression] = e match {
```

Comment:
Eventually, the solution is ContProp on high firrtl. I'll fix both `getOrigin` and this with it.